File: frontend/types/websocket.ts
// WebSocket event type definitions

import type { User, Message } from './session';

// Client → Server Events
export interface SendMessageEvent {
  event: 'send_message';
  data: {
    user_name: string;
    message: string;
  };
}

export interface JoinSessionEvent {
  event: 'join_session';
  data: {
    user_id: string;
    name: string;
  };
}

export interface LeaveSessionEvent {
  event: 'leave_session';
  data: {};
}

export interface TypingEvent {
  event: 'typing';
  data: {
    user_name: string;
  };
}

export type ClientEvent = SendMessageEvent | JoinSessionEvent | LeaveSessionEvent | TypingEvent;

// Server → Client Events
export interface UserJoinedEvent {
  event: 'user_joined';
  data: User;
}

export interface UserLeftEvent {
  event: 'user_left';
  data: User;
}

export interface MessageReceivedEvent {
  event: 'message_received';
  data: {
    user_name: string;
    message: string;
    timestamp: string;
  };
}

export interface SpiritThinkingEvent {
  event: 'spirit_thinking';
  data: {};
}

export interface SpiritResponseEvent {
  event: 'spirit_response';
  data: {
    message: string;
    letter_timings: number[];
    timestamp: string;
  };
}

export interface ErrorEvent {
  event: 'error';
  data: {
    message: string;
    code?: string;
  };
}

export type ServerEvent = 
  | UserJoinedEvent 
  | UserLeftEvent 
  | MessageReceivedEvent 
  | SpiritThinkingEvent 
  | SpiritResponseEvent 
  | ErrorEvent;

// WebSocket message wrapper
export interface WebSocketMessage<T = any> {
  event: string;
  data: T;
  timestamp?: string;
}

File: backend/app/schemas/websocket.py
"""WebSocket event schemas for real-time communication."""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class EventType(str, Enum):
    """WebSocket event types."""

    # Client → Server
    SEND_MESSAGE = "send_message"
    TYPING = "typing"

    # Server → Client
    USER_JOINED = "user_joined"
    USER_LEFT = "user_left"
    MESSAGE_RECEIVED = "message_received"
    SPIRIT_THINKING = "spirit_thinking"
    SPIRIT_RESPONSE = "spirit_response"
    ERROR = "error"


class WSEvent(BaseModel):
    """Base WebSocket event structure."""

    event: EventType = Field(..., description="Event type")
    data: Dict[str, Any] = Field(default_factory=dict, description="Event data")
    timestamp: datetime = Field(default_factory=datetime.utcnow, description="Event timestamp")


class UserInfo(BaseModel):
    """User information for WebSocket events."""

    id: str = Field(..., description="User ID")
    name: str = Field(..., description="User name")
    joined_at: datetime = Field(default_factory=datetime.utcnow, description="Join timestamp")


class UserJoinedEvent(BaseModel):
    """Event data for user_joined event."""

    event: str = Field(default="user_joined", description="Event type")
    data: UserInfo = Field(..., description="User information")
    timestamp: datetime = Field(default_factory=datetime.utcnow, description="Event timestamp")


class UserLeftEvent(BaseModel):
    """Event data for user_left event."""

    event: str = Field(default="user_left", description="Event type")
    data: UserInfo = Field(..., description="User information")
    timestamp: datetime = Field(default_factory=datetime.utcnow, description="Event timestamp")


class MessageReceivedEvent(BaseModel):
    """Event data for message_received event."""

    event: str = Field(default="message_received", description="Event type")
    data: Dict[str, Any] = Field(..., description="Message data")
    timestamp: datetime = Field(default_factory=datetime.utcnow, description="Event timestamp")


class SpiritThinkingEvent(BaseModel):
    """Event data for spirit_thinking event."""

    event: str = Field(default="spirit_thinking", description="Event type")
    data: Dict[str, Any] = Field(default_factory=dict, description="Event data")
    timestamp: datetime = Field(default_factory=datetime.utcnow, description="Event timestamp")


class SpiritResponseEvent(BaseModel):
    """Event data for spirit_response event."""

    event: str = Field(default="spirit_response", description="Event type")
    data: Dict[str, Any] = Field(..., description="Spirit response data")
    timestamp: datetime = Field(default_factory=datetime.utcnow, description="Event timestamp")


class ErrorEvent(BaseModel):
    """Event data for error event."""

    event: str = Field(default="error", description="Event type")
    data: Dict[str, str] = Field(..., description="Error information")
    timestamp: datetime = Field(default_factory=datetime.utcnow, description="Event timestamp")
